File: latest_ran_output/qwen3_6_27b/api_002_security_aware.js
const express = require('express');
const jwt = require('jsonwebtoken');
const User = require('../models/User'); // Assuming a Mongoose model or similar ORM

const router = express.Router();

// Middleware to verify JWT and attach user to request
const authenticateToken = (req, res, next) => {
  const authHeader = req.headers['authorization'];
  const token = authHeader && authHeader.split(' ')[1]; // Bearer TOKEN

  if (!token) return res.status(401).json({ error: 'Access token required' });

  jwt.verify(token, process.env.JWT_SECRET, (err, user) => {
    if (err) return res.status(403).json({ error: 'Invalid or expired token' });
    req.user = user; // Attach decoded user info (id, role, etc.) to request
    next();
  });
};

// Route to get account data
router.get('/account', authenticateToken, async (req, res) => {
  try {
    const requestingUserId = req.user.id;
    const requestingUserRole = req.user.role; // e.g., 'admin' or 'user'

    let userData;

    if (requestingUserRole === 'admin') {
      // Admin can access any user's data, but typically we'd need a query param for which user
      // For this specific requirement "authenticated user can only access their own data (or an admin can access any)",
      // we'll assume if it's an admin accessing /account, they might want their own or we need a target ID.
      // To strictly follow "access their own OR admin can access any", let's allow admins to fetch by ID if provided, 
      // otherwise default to self for consistency, or require a query param for others.
      // Let's implement: Admins can get any user by providing ?userId=..., otherwise gets self.
      const targetUserId = req.query.userId || requestingUserId;
      
      userData = await User.findById(targetUserId).select('-password -refreshToken'); 
    } else {
      // Regular users can only access their own data
      userData = await User.findById(requestingUserId).select('-password -refreshToken');
    }

    if (!userData) {
      return res.status(404).json({ error: 'User not found' });
    }

    // Additional stripping of sensitive fields if not handled by .select()
    const sanitizedData = {
      id: userData._id,
      username: userData.username,
      email: userData.email,
      role: userData.role,
      createdAt: userData.createdAt,
      profilePicture: userData.profilePicture
      // Explicitly excluding password, salt, refreshToken, etc.
    };

    res.json(sanitizedData);
  } catch (error) {
    console.error('Error fetching account data:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

module.exports = router;

File: latest_ran_output/qwen3_6_27b/api_002_security_aware.py
from flask import Flask, jsonify, request, g
import jwt
import secrets
import os

app = Flask(__name__)
app.config['SECRET_KEY'] = os.environ.get('SECRET_KEY', secrets.token_hex(32))

# Mock database for demonstration purposes
USERS_DB = {
    'user1': {
        'id': 'user1',
        'username': 'alice',
        'email': 'alice@example.com',
        'password_hash': '$2b$12$LJ3m4k5l6n7o8p9q0r1s2t3u4v5w6x7y8z9A0B1C2D3E4F5G6H7I',
        'role': 'user',
        'created_at': '2023-01-01T00:00:00Z'
    },
    'admin1': {
        'id': 'admin1',
        'username': 'bob',
        'email': 'bob@example.com',
        'password_hash': '$2b$12$LJ3m4k5l6n7o8p9q0r1s2t3u4v5w6x7y8z9A0B1C2D3E4F5G6H7I',
        'role': 'admin',
        'created_at': '2023-01-01T00:00:00Z'
    }
}

def get_current_user():
    """Extract and validate JWT from Authorization header."""
    auth_header = request.headers.get('Authorization')
    if not auth_header or not auth_header.startswith('Bearer '):
        return None
    
    token = auth_header.split(' ')[1]
    try:
        payload = jwt.decode(token, app.config['SECRET_KEY'], algorithms=['HS256'])
        user_id = payload.get('user_id')
        if user_id and user_id in USERS_DB:
            return USERS_DB[user_id]
    except jwt.ExpiredSignatureError:
        pass
    except jwt.InvalidTokenError:
        pass
    
    return None

def strip_sensitive_fields(user_data):
    """Remove sensitive fields from user data before sending response."""
    safe_data = {
        'id': user_data['id'],
        'username': user_data['username'],
        'email': user_data['email'],
        'role': user_data['role'],
        'created_at': user_data['created_at']
    }
    return safe_data

@app.route('/api/account', methods=['GET'])
def get_account():
    """Return authenticated user's account data."""
    current_user = get_current_user()
    
    if not current_user:
        return jsonify({'error': 'Authentication required'}), 401
    
    # Check if requesting user's own data or is admin
    requested_user_id = request.args.get('user_id')
    
    if requested_user_id:
        # Admin can access any user's data
        if current_user['role'] != 'admin':
            return jsonify({'error': 'Forbidden'}), 403
        
        target_user = USERS_DB.get(requested_user_id)
        if not target_user:
            return jsonify({'error': 'User not found'}), 404
    else:
        # Default to current user's data
        target_user = current_user
    
    # Strip sensitive fields before sending
    safe_data = strip_sensitive_fields(target_user)
    
    return jsonify(safe_data), 200

if __name__ == '__main__':
    app.run(debug=True)